extension/property_tree/include/property_tree/property_tree.hpp

```diff
@@ -0,0 +1,228 @@
+    {
+        auto sep = path.find(".");
+        if (sep == std::string::npos) {
+            return children_.at(key_map_.at(path));
+        }
+        return children_.at(key_map_.at(path.substr(0, sep)))
+            .get_child(path.substr(sep + 1));
+    }
+
+    // Get the index i children
+    pnode_s<data_type>& get_child(int i) { return children_.at(i); }
+
+    // Check the status
+    bool is(status_t s) const { return this->get_status() == s; }
+
+    int get_size() const { return children_.size(); }
+
+    status_t get_status() const { return status_; }
+
+    std::string get_name() const { return name_; }
+
+    // Only allow change the status from empty
+    void update_status(status_t status)
+    {
+        if (status_ != status_t::empty && status_ != status) {
+            throw std::runtime_error("Can not change the status");
+        }
+        if (status == status_t::empty) {
+            throw std::runtime_error("Can not clear");
+        }
+        status_ = status;
+    }
+
+    void set(const data_type& data)
+    {
+        update_status(status_t::object);
+        data_ = data;
+    }
+
+    void insert(const std::string& name, const data_type& data)
+    {
+        this->update_status(status_t::object_list);
+        this->insert_item(name, data);
+    }
+
+    // insert for array
+    void insert(const data_type& data)
+    {
+        this->update_status(status_t::array);
+        int size = key_map_.size();
+        std::string name = "array_" + std::to_string(size);
+        this->insert_item(name, data);
+    }
+
+    void insert(std::initializer_list<data_type> data)
+    {
+        this->update_status(status_t::array);
+        for (auto item : data) {
+            this->insert(item);
+        }
+    }
+
+    void allocate(const std::string& name)
+    {
+        this->update_status(status_t::object_list);
+        this->insert_item(name);
+    }
+
+    void allocate_array(int num)
+    {
+        this->update_status(status_t::array);
+        auto size = children_.size();
+        for (int i = 0; i < num; i++) {
+            std::string name = "array_" + std::to_string(size + i);
+            this->insert_item(name);
+        }
+    }
+
+private:
+    // It's for insert item, please check the status before it
+    void insert_item(const std::string& name, const data_type& data)
+    {
+        int size = key_map_.size();
+        auto p = key_map_.emplace(name, size);
+        if (!p.second) {
+            throw std::runtime_error("Have the same key");
+        }
+        children_.emplace_back(pnode_s{name, data});
+    }
+
+    // insert empty
+    void insert_item(const std::string& name)
+    {
+        int size = key_map_.size();
+        auto p = key_map_.emplace(name, size);
+        if (!p.second) {
+            throw std::runtime_error("Have the same key");
+        }
+        children_.emplace_back(pnode_s{name});
+    }
+
+    std::string name_;
+    std::deque<pnode_s<data_type>> children_;
```

Comment:
you only seem to be pushing to the back, and accessing by index, that works better in a vector. Why do we need to use data_type as a template parameter? Can't we set it to `data_type` globally?
```suggestion
    std::vector<pnode_s<data_type>> children_;
```